### DISCUSSION TITLE
Composite key strategies

### DISCUSSION THREAD

**Discussion by thesmartwon** [tinyplex/tinybase#195] [General]
What is the best pattern for composite keys? Is it:
```ts
const schema = {
  // Incrementing rowId
  table: { key1: { type: "string" }, key2: { "type": "string" }, ... },
};
```
OR
```ts
const schema = {
  // `${key1}/${key2}` rowId
  table: { ... }
};
```

I don't have enough experience with the library yet to be able to evaluate the differences. I imagine the first interops better with existing SQL databases and queries.

**Comment by waynesbrain** [comment]
There is no support for composite keys in TinyBase as far as I know. Each row id of a TinyBase table is a single unique string field which does not get defined in the table schema.

If you use the store's `setRow` method instead of `addRow`, then you can set your id to any string. Otherwise, `addRow` will generate an id for you. Currently it generates the ids by simply incrementing an integer and converting to string, but there is not guarantee on how it will generate the id string.

Also, if you do set `${key1}/${key2}` as the id of a table then you will have to split that id yourself in order to lookup `key1` and `key2` in the related tables. For example:

```ts
const compositeRowId = `${key1}/${key2}`; // key from table 2
const parts = compositeRowId.split("/");
const related1 = store.getRow('table1', parts[0]);
const related2 = store.getRow('table1', parts[1]);
```

**Comment by waynesbrain** [comment]
Check out my code here to see how I created a nested structure - https://github.com/tinyplex/tinybase/discussions/180
